### DISCUSSION TITLE
Possible tie-breaking of equivalent Clingo models

### DISCUSSION THREAD

**Discussion by joshessman-llnl** [spack/spack#22242] [General]
Currently Spack's `PyClingoDriver` selects a model of minimum cost with `min`, but there does not appear to be any explicit logic that checks for multiple models of equivalent cost (in which case `min` will select the first, but I am not sure that Clingo will produce a list with the same order each time).

Is it possible that this could lead to nondeterministic behavior where different concretizations could be chosen for a given spec?

Relevant snippet:
https://github.com/spack/spack/blob/d0610f7a39c44e919f9bdd147f1cdb351cb6931b/lib/spack/spack/solver/asp.py#L361

Tag: @becker33

**Comment by alalazo** [comment]
> Is it possible that this could lead to nondeterministic behavior where different concretizations could be chosen for a given spec?

This should not happen frequently, but it's possible that two models have the same cost and which one gets picked depend on the run.

**Comment by becker33** [comment]
@alalazo do you have any ideas about how we could eliminate that sort of nondeterminism? I should think that a couple more low-priority minimization criteria should be enough -- for variants, for example, we could have a secondary target to minimize the combined index of the non-default variants in a flattened list of nodes/variants.
